### PR TITLE
Set underlying console QTextEdit widget as proxy for QtConsole focus

### PR DESCRIPTION
**Pull request by aganders3**
This is meant to be a precursor to fixing https://github.com/napari/napari/issues/5166, allowing napari to programmatically set the keyboard focus on the console.

This is draft for now as I intended to open this in my own fork and need to fix the test failures in CI. Sorry!

### CODE REVIEW

File: napari_console/_tests/test_qt_console.py [hunk 1]
```diff
@@ -46,3 +46,26 @@ def mock_get_ipython():
         console = QtConsole(viewer)
         qtbot.addWidget(console)
         assert console.kernel_client is None
+
+
+def test_console_focus_proxy(qtbot, make_test_viewer):
+    """Test setting/clearing focus on a QtConsole sets/clears focus on the underlying QTextEdit"""
+    viewer = make_test_viewer()
+    console = QtConsole(viewer)
+    qtbot.addWidget(console)
```
Comment (Czaki) [inline napari_console/_tests/test_qt_console.py:55]: why create a separate console instead of `viewer.window._qt_viewer.console`?

Comment (aganders3) [reply napari_console/_tests/test_qt_console.py:55]: Good question - I just copied that part of the test from the above tests but can try this instead.

**Comment by Czaki** [comment]
the test failure is connected with imageio 2.22.1 release You need to block this version or wait until the next imageio release (as the bugfix is already merged in imageio).

**Comment by aganders3** [comment]
> the test failure is connected with imageio 2.22.1 release You need to block this version or wait until the next imageio release (as the bugfix is already merged in imageio).

Thanks - this saved me some digging and worked well, but now the tests fail for a reason I'm struggling with.

It seems to me likely related to https://github.com/pytest-dev/pytest-qt/issues/206 but I'm going to close this PR until I can figure it out in my own fork.

**Comment by Czaki** [comment]
I'm familiar with this type of problem. Point it out, and maybe I could provide you a proper solution.

**Comment by aganders3** [comment]
Thank you! I managed to get it working in my fork but of course appreciate your input on the changes. I'll reopen this so you can look here.

File: .github/workflows/test_and_deploy.yml [hunk 2]
```diff
@@ -53,10 +59,17 @@ jobs:
...
-        with:
-          run: python -m tox -vv
+        run: |
+          if [ "$RUNNER_OS" == "Linux" ] ; then
+            Xvfb :99 -screen 0 1024x768x24 & sleep 3
+            herbstluftwm & sleep 1
+            python -m tox -vv
+          else
+            python -m tox -vv
+          fi
```
Comment (Czaki) [inline .github/workflows/test_and_deploy.yml:69]: why did you need to change from `xvfb-action` to this code? In napari we use this action without any problem. It will be nice to see error.

Comment (aganders3) [reply .github/workflows/test_and_deploy.yml:69]: I couldn't see a way to run the window manager on the correct display with the action, because it uses `xvfb-run` and that manages the `xvfb` process for you.

Comment (Czaki) [reply .github/workflows/test_and_deploy.yml:69]: But some test fails?

Comment (aganders3) [reply .github/workflows/test_and_deploy.yml:69]: Oh I guess it was just a syntax error with that OS detection logic I put in. I will try instead to put that logic in the action to see if it works. Something like this:

```
      - name: Test with tox (Xvfb)
        if: runner.os == "Linux"
        uses: GabrielBB/xvfb-action@v1
        with:
          run: |
            herbstluftwm & sleep 1
            python -m tox -vv

      - name: Test with tox
        if: runner.os != "Linux"
        run: python -m tox -vv
        env:
          PLATFORM: ${{ matrix.platform }}
          PYVISTA_OFF_SCREEN: True
```

Comment (aganders3) [reply .github/workflows/test_and_deploy.yml:69]: It just hangs: https://github.com/aganders3/napari-console/actions/runs/3191198869/jobs/5207212911

Maybe the wm could be run in the background by tox or even pytest (in a fixture, for example)?

Comment (aganders3) [reply .github/workflows/test_and_deploy.yml]: > Maybe the wm could be run in the background by tox or even pytest (in a fixture, for example)?

This seemed to work (see https://github.com/napari/napari-console/pull/23/commits/23fc6cd90ea56ce7a0b088fcfb596d2975bd2eaa) and here's the passing test run: https://github.com/aganders3/napari-console/actions/runs/3191679449/jobs/5208278101

It could use a little cleanup but let me know what you think. I appreciate your feedback on the approach.

Comment (Czaki) [reply .github/workflows/test_and_deploy.yml]: it looks ok

**Review by Czaki — APPROVED** [review]
I like changes in code, but I still need to verify if there is a need to use `herbstluftwm`

**Comment by aganders3** [comment]
Thanks - my impression is that any minimal WM should work and I only selected `herbstluftwm` because it was suggested by the `pytest-qt` docs. I did very briefly try using `twm` but it was not immediately working. Let me know if there's anything else you'd like me to try. If there's a way to do this without running any WM at all that would be great!

**Comment by Czaki** [comment]
When I locally test, it is enough to set the environment available to `QT_QPA_PLATFORM=offscreen`and add `QT_QPA_PLATFORM` to passenv section in tox. 

This should allow us to not use xvfb at all. But I'm not sure which variant is better.

**Comment by aganders3** [comment]
Oh neat thanks - I was not aware of `QT_QPA_PLATFORM`.

**Comment by psobolewskiPhD** [comment]
This PR along with https://github.com/napari/napari/issues/5166 works wonderfully as advertised. I also tested +/- each of the PRs and there are no issues with one or the other is missing.

**Review by Czaki — APPROVED** [review]

File: napari_console/_tests/test_qt_console.py [hunk 3]
```diff
@@ -46,3 +46,30 @@ def mock_get_ipython():
         console = QtConsole(viewer)
         qtbot.addWidget(console)
         assert console.kernel_client is None
+
+
+def test_console_focus_proxy(qtbot, make_test_viewer, linux_wm):
```
Comment (Czaki) [inline napari_console/_tests/test_qt_console.py:51]:
```suggestion
def test_console_focus_proxy(qtbot, make_test_viewer, linux_wm_local):
```

File: napari_console/conftest.py [hunk 4]
```diff
@@ -0,0 +1,37 @@
+import os
+import platform
+import subprocess
+import time
+
+import pytest
+
+
+@pytest.fixture
+def linux_wm():
```
Comment (Czaki) [inline napari_console/conftest.py:10]:
```suggestion
def linux_wm_local():
```

**Review by Czaki — COMMENTED** [review]
To avoid name collision on next napari release

**Comment by aganders3** [comment]
Thanks again for all the review effort and suggestions!
